### ISSUE TITLE
React button not showing icon

### ISSUE THREAD

**Issue by StephaneAdeso** [microsoft/vscode-webview-ui-toolkit#508]
Report this ad
0

I am developing an extension for vscode with React and using the WebUi Toolkit library for components. I want to add a "save" icon to my button.

I have followed the documentation to add my button in react and add an icon.
So the resulting code is:
```tsx
import { VSCodeButton } from '@vscode/webview-ui-toolkit/react';    

<VSCodeButton >
    Save
    <span slot="start" className="codicon codicon-save"></span>
</VSCodeButton>
```
But the icon is not showing. How to add the icon to the button?

![image](https://github.com/microsoft/vscode-webview-ui-toolkit/assets/55595063/9975509b-8dbc-405c-be4d-6644728dcb15)

**Comment by hawkticehurst** [comment]
Hey @StephaneAdeso! Out of curiosity what does your content security policy look like in your extension? This looks like a case where the icon is being blocked from rendering.

As an example, this CSP should probably look something like this: https://github.com/microsoft/vscode-webview-ui-toolkit-samples/blob/main/frameworks/component-gallery-react/src/panels/ComponentGalleryPanel.ts#L122

In particular, the `font-src` CSP rule should allow the `codicons.ttf` (required to render Codicons) file to be rendered.

**Comment by StephaneAdeso** [comment]
Thank you very much for answering. After several days without finding a solution, I finally decided to use an icon library. I must also say that you are right. I did not have any content security policy established. As soon as I have some time I'll try it. Thank you so much.

**Comment by hawkticehurst** [comment]
Of course! Please let me know how it goes so I can know if this issue is resolved and I can close it :)

**Comment by StephaneAdeso** [comment]
Sorry, i was unable to find a correct CSP configuration that works. All variations i tried throw errors so i removed it.

**Comment by hawkticehurst** [comment]
Hmm okay, in that case, could you create/provide a link to a GitHub repo with a small reproduction of the issue so I can see if I can figure out what the problem might be?

**Comment by maninak** [comment]
I'm using vue and I have the same issue.

My CSP is the following:

```hml
      <meta
        http-equiv="Content-Security-Policy"
        content="
          default-src 'none';
          object-src 'none';
          base-uri 'none';
          style-src ${allowedSource} 'unsafe-inline';
          img-src ${allowedSource} https: data:;
          script-src 'strict-dynamic' 'nonce-${nonce}' 'unsafe-inline' https:;
        "
      >
```

But I don't think it's a CSP issue because I get no CSP-related issues logged in the webview's dev tools' console and even if I delete the whole CSP I still have that issue.

Repro steps

1. $ git clone git@github.com:cytechmobile/radicle-vscode-extension.git
2. $ git checkout 4ee3787d591a1612242b43b663bd7167fbee9fad
3. $ npm run dev
4. Ctrl + Shift + P (to open Cmd Palette) and select command "Open dat webview"

you should be seeing this

https://github.com/microsoft/vscode-webview-ui-toolkit/assets/7542054/146d0118-aea1-4536-8773-6426f38f5d1f

The related code can be found under `src/webviews/src/components/Counter.vue` and looks like this

```html
    <vscode-button appearance="icon" @click="showInfoNotifWithCount">
      <span class="codicon codicon-megaphone"></span>
    </vscode-button>
```

**Comment by r3m0t** [comment]
You are missing the `font-src` directive, so `default-src 'none'` is taking precedence.

**Comment by maninak** [comment]
@r3m0t if you could share a diff so that I can better understand where I'm missing it that would help a lot, thank you!

**Comment by hawkticehurst** [comment]
Hey @maninak! Yes, as @r3m0t said this looks like a case of not including the `font-src` CSP directive  (thank you for chiming in @r3m0t btw 😊).

You would add the font-src CSP to the `content` attribute in the meta tag you wrote out above. Something like this:

```html
<meta 
  http-equiv="Content-Security-Policy" 
  content="
    default-src 'none'; 
    script-src 'nonce-${nonce}';
    style-src ${webview.cspSource}; 
    font-src ${webview.cspSource};   // <-- here!
">
```

The above example is specifically pulled from our [component-gallery sample extension](https://github.com/microsoft/vscode-webview-ui-toolkit-samples/blob/main/default/component-gallery/src/panels/ComponentGalleryPanel.ts#L127). Feel free to look there for more context.

Also, the [content security policy docs](https://code.visualstudio.com/api/extension-guides/webview#content-security-policy) live in Webview API VS Code documentation since CSP is a topic that applies to webview extension development in general, not just in the Webview UI Toolkit.

Hope that helps, but please let me know if there are any other questions you have!

**Comment by maninak** [comment]
Thank you both for the quick respons and especially Hawk for the detailed hints.

I've changed my CSP to be

```html

      <meta
        http-equiv="Content-Security-Policy"
        content="
          default-src 'none';
          object-src 'none';
          base-uri 'none';
          style-src ${allowedSource} 'unsafe-inline';
          img-src ${allowedSource} https: data:;
          script-src 'strict-dynamic' 'nonce-${nonce}' 'unsafe-inline' https:;
          font-src ${allowedSource};
        "
      >
```

The above seemed to have no effect. Noteworthy is that the chromium devtools report no CSP warnings or errors! Neither did they before nor do they now after adding the font-src directive.

**Comment by r3m0t** [comment]
What does the Network tab show after filtering for Font?

Is the codicon.css stylesheet referenced in your project?

**Comment by maninak** [comment]
Hey @r3m0t, after reading your comment I didn't know to which `codicon.css` you were referring to. There's no mention of it on the Readme nor the Getting Started Guide.

After checking the [link](https://github.com/microsoft/vscode-webview-ui-toolkit-samples/blob/main/default/component-gallery/src/panels/ComponentGalleryPanel.ts#L113) @hawkticehurst shared above I found that in that sample there's an explicit import to `codicon.css` which also imports `codicon.ttf` and _both are local in the repo_.

I copy-pasted those into my vue repo's `assets/` folder and imported the css like so inside my `App.vue`.

```html
<style>
@import './assets/codicon.css';
</style>
```

This has definitely moved me closer but now I'm seeing this in the console and the network tab
![image](https://github.com/microsoft/vscode-webview-ui-toolkit/assets/7542054/187d8dea-0099-43dc-8fa6-766b8758b003)
![image](https://github.com/microsoft/vscode-webview-ui-toolkit/assets/7542054/e6820356-d0fc-4746-938c-d09c54b1950c)

**Comment by maninak** [comment]
OK, so after copying the css and ttf file under `assets/` and importing it in my `App.vue` what was left was setting `base: ''` in the vue app's `vite.config.ts` which, for the sake of completion, now looks like this:

```ts
import { fileURLToPath, URL } from 'node:url'

import { defineConfig } from 'vite'
import vue from '@vitejs/plugin-vue'

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [
    vue(),
  ],
  resolve: {
    alias: {
      '@': fileURLToPath(new URL('./src', import.meta.url)),
      'lib': fileURLToPath(new URL('../../lib', import.meta.url)),
      'utils': fileURLToPath(new URL('../utils', import.meta.url)),
    }
  },
  build: {
    rollupOptions: {
      external: ['vscode'],
      // produce predictable filenames without cache-busting SHA suffix
      output: {
        entryFileNames: `assets/[name].js`,
        chunkFileNames: `assets/[name].js`,
        assetFileNames: `assets/[name].[ext]`,
      },
    },
  },
  base: '', // <-- this fixed the problem with the resolved path to codicons.ttf :man_shrugging: 
})
```

Of course, this is more specific to my use case, but I'd argue that it is tangential to the scope of this issue, since I wouldn't have to deal with all of this if the UI library would make sure to import its necessary assets by itself.
